nixos/modules/security/permissions-wrappers/default.nix

```diff
@@ -0,0 +1,198 @@
+{ config, lib, pkgs, ... }:
+let
+
+  inherit (config.security) run-permissionsWrapperDir permissionsWrapperDir;
+
+  isNotNull = v: if v != null then true else false;
+
+  cfg = config.security.permissionsWrappers;
+
+  setcapWrappers = import ./setcap-wrapper-drv.nix {
+    inherit config lib pkgs;
+  };
+
+  setuidWrappers = import ./setuid-wrapper-drv.nix {
+    inherit config lib pkgs;
+  };
+
+  ###### Activation script for the setcap wrappers
+  configureSetcapWrapper =
+    { program
+    , capabilities
+    , source ? null
+    , owner  ? "nobody"
+    , group  ? "nogroup"
+    }: ''
+      cp ${setcapWrappers}/bin/${program}.wrapper $permissionsWrapperDir/${program}
+
+      # Prevent races
+      chmod 0000 $permissionsWrapperDir/${program}
+      chown ${owner}.${group} $permissionsWrapperDir/${program}
+
+      # Set desired capabilities on the file plus cap_setpcap so
+      # the wrapper program can elevate the capabilities set on
+      # its file into the Ambient set.
+      #
+      # Only set the capabilities though if we're being told to
+      # do so.
+      ${pkgs.libcap.out}/bin/setcap "cap_setpcap,${capabilities}" $permissionsWrapperDir/${program}
+
+      # Set the executable bit
+      chmod u+rx,g+x,o+x $permissionsWrapperDir/${program}
+    '';
+
+  ###### Activation script for the setuid wrappers
+  configureSetuidWrapper =
+    { program
+    , source ? null
+    , owner  ? "nobody"
+    , group  ? "nogroup"
+    , setuid ? false
+    , setgid ? false
+    , permissions ? "u+rx,g+x,o+x"
+    }: ''
+      cp ${setuidWrappers}/bin/${program}.wrapper $permissionsWrapperDir/${program}
+
+      # Prevent races
+      chmod 0000 $permissionsWrapperDir/${program}
+      chown ${owner}.${group} $permissionsWrapperDir/${program}
+
+      chmod "u${if setuid then "+" else "-"}s,g${if setgid then "+" else "-"}s,${permissions}" $permissionsWrapperDir/${program}
+    '';
+in
+{
+
+  ###### interface
+
+  options = {
+    security.permissionsWrappers.setcap = lib.mkOption {
+      type    = lib.types.listOf lib.types.attrs;
+      default = [];
+      example =
+        [ { program = "ping";
+            source  = "${pkgs.iputils.out}/bin/ping";
+            owner   = "nobody";
+            group   = "nogroup";
+            capabilities = "cap_net_raw+ep";
+          }
+        ];
+      description = ''
+        This option sets capabilities on a wrapper program that
+        propagates those capabilities down to the wrapped, real
+        program.
+
+        The `program` attribute is the name of the program to be
```

Comment:
the docs we have to use `<literal></literal>`, backticks won't work (I've failed this way too often)